### PR TITLE
 [AIRFLOW-4535] Break jobs.py into multiple files

### PR DESCRIPTION
**Pull request by milton0825**
Make sure you have checked _all_ steps below.

### Jira

- [X] My PR addresses the following [Airflow Jira](https://issues.apache.org/jira/browse/AIRFLOW/) issues and references them in the PR title. For example, "\[AIRFLOW-XXX\] My Airflow PR"
  - https://issues.apache.org/jira/browse/AIRFLOW-4535

### Description

- [X] Here are some details about my PR, including screenshots of any UI changes:

Split `jobs.py` and `test_jobs.py` into multiple files to improve readability.

### Tests

- [X] My PR adds the following unit tests __OR__ does not need testing for this extremely good reason:

No change to logic.

### Commits

- [X] My commits all reference Jira issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
  1. Subject is separated from body by a blank line
  1. Subject is limited to 50 characters (not including Jira issue reference)
  1. Subject does not end with a period
  1. Subject uses the imperative mood ("add", not "adding")
  1. Body wraps at 72 characters
  1. Body explains "what" and "why", not "how"

### Documentation

- [X] In case of new functionality, my PR adds documentation that describes how to use it.
  - All the public functions and the classes in the PR contain docstrings that explain what it does
  - If you implement backwards incompatible changes, please leave a note in the [Updating.md](https://github.com/apache/airflow/blob/master/UPDATING.md) so we can assign it to a appropriate release

### Code Quality

- [X] Passes `flake8`

### CODE REVIEW

**Comment by BasPH** [comment]
We had the same idea 😄 https://github.com/apache/airflow/pull/5300 & https://issues.apache.org/jira/browse/AIRFLOW-4536

**Comment by milton0825** [comment]
@BasPH uhhh... what should we do now... Our jira happens to be consecutive..

**Comment by milton0825** [comment]
@BasPH I think I got the fix.

**Comment by BasPH** [comment]
I closed my PR, let's continue this one.

File: airflow/jobs/base_job.py [hunk 1]
```diff
@@ -0,0 +1,279 @@
...
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+#
+
+import getpass
+from time import sleep
+from sqlalchemy import (Column, Index, Integer, String, and_, or_)
```
Comment (BasPH) [inline airflow/jobs/base_job.py]: Could you add a new line here? `sqlalchemy` is a 3rd party library and `time` & `getpass` are native Python.

File: airflow/jobs/base_job.py [hunk 2]
```diff
@@ -0,0 +1,279 @@
...
+from airflow.stats import Stats
+from airflow.utils import helpers, timezone
+from airflow.utils.db import create_session, provide_session
+from airflow.utils.log.logging_mixin import LoggingMixin
+from airflow.utils.net import get_hostname
+from airflow.utils.sqlalchemy import UtcDateTime
+from airflow.utils.state import State
+
+Base = models.base.Base  # type: Any
+ID_LEN = models.base.ID_LEN
```
Comment (BasPH) [inline airflow/jobs/base_job.py]: Base & ID_LEN are also exposed in `/airflow/models/__init__.py`, you could simply import as `from airflow.models import Base, ID_LEN`.

File: airflow/jobs/local_task_job.py [hunk 3]
```diff
@@ -0,0 +1,177 @@
...
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+#
+
+import os
+import signal
+import time
+from sqlalchemy.exc import OperationalError
```
Comment (BasPH) [inline airflow/jobs/local_task_job.py:25]: Also newline here

File: airflow/jobs/local_task_job.py [hunk 4]
```diff
@@ -0,0 +1,177 @@
...
+)
+from airflow.stats import Stats
+from airflow.task.task_runner import get_task_runner
+from airflow.utils.db import provide_session
+from airflow.utils.net import get_hostname
+from airflow.jobs.base_job import BaseJob
+from airflow.utils.state import State
+
+Base = models.base.Base  # type: Any
+ID_LEN = models.base.ID_LEN
```
Comment (BasPH) [inline airflow/jobs/local_task_job.py]: And same comment as in base_job here

File: airflow/jobs/scheduler_job.py [hunk 5]
```diff
@@ -16,292 +16,49 @@
...
 import time
-from collections import OrderedDict, defaultdict
+from collections import defaultdict
 from datetime import timedelta
 from time import sleep
 
 import six
-from sqlalchemy import (Column, Index, Integer, String, and_, func, not_, or_)
-from sqlalchemy.exc import OperationalError
+from sqlalchemy import (and_, func, not_, or_)
```
Comment (BasPH) [inline airflow/jobs/scheduler_job.py]: no need for the brackets here

File: airflow/jobs/scheduler_job.py [hunk 6]
```diff
@@ -1059,6 +816,7 @@ def _find_executable_task_instances(self, simple_dag_bag, states, session=None):
         :type states: tuple[airflow.utils.state.State]
         :return: list[airflow.models.TaskInstance]
         """
+        from airflow.jobs.backfill_job import BackfillJob
```
Comment (BasPH) [inline airflow/jobs/scheduler_job.py]: Perhaps add a comment `# Avoid circular import`?

File: tests/jobs/test_local_task_job.py [hunk 7]
```diff
@@ -0,0 +1,203 @@
...
+from airflow.utils import timezone
+from airflow.utils.net import get_hostname
+from airflow.utils.state import State
+from tests.compat import patch
+from tests.core import TEST_DAG_FOLDER
+from tests.test_utils.db import clear_db_runs
+
+configuration.load_test_config()
+
+logger = logging.getLogger(__name__)
```
Comment (BasPH) [inline tests/jobs/test_local_task_job.py]: I don't believe the `logger` is used in this script, is that necessary here?

File: tests/jobs/test_base_job.py [hunk 8]
```diff
@@ -0,0 +1,70 @@
...
+import logging
+import unittest
+
+from airflow import configuration
+from airflow.jobs import BaseJob
+from airflow.utils.state import State
+
+configuration.load_test_config()
+
+logger = logging.getLogger(__name__)
```
Comment (BasPH) [inline tests/jobs/test_base_job.py]: I don't believe the `logger` is used in this script, is that necessary here?

File: tests/jobs/test_scheduler_job.py [hunk 9]
```diff
@@ -79,1622 +68,7 @@
...
-        with patch.object(BaseTaskRunner, 'start', return_value=None) as mock_method:
-            job1.run()
-            mock_method.assert_not_called()
-
-        ti = dr.get_task_instance(task_id=task.task_id, session=session)
-        self.assertEqual(ti.pid, 1)
-        self.assertEqual(ti.state, State.RUNNING)
-
-        session.close()
+    os.path.dirname(os.path.dirname(os.path.realpath(__file__))), 'dags')
```
Comment (BasPH) [inline tests/jobs/test_scheduler_job.py]: In other test scripts we do `from tests.core import TEST_DAG_FOLDER`, shall we stick to that method everywhere?

File: tests/www/api/experimental/test_dag_runs_endpoint.py [hunk 10]
```diff
@@ -35,6 +35,9 @@ def setUpClass(cls):
         session.query(DagRun).delete()
         session.commit()
         session.close()
+        dagbag = DagBag(include_examples=True)
```
Comment (BasPH) [inline tests/www/api/experimental/test_dag_runs_endpoint.py:38]: I don't understand this change, I assume it was necessary to load the example DAGs? Did it not do that before this change?

File: airflow/jobs/base_job.py (hunk 1 above)
Comment (milton0825) [reply airflow/jobs/base_job.py]: Fixed.

File: airflow/jobs/base_job.py (hunk 2 above)
Comment (milton0825) [reply airflow/jobs/base_job.py]: Fixed.

File: airflow/jobs/local_task_job.py (hunk 3 above)
Comment (milton0825) [reply airflow/jobs/local_task_job.py:25]: added.

Comment (milton0825) [reply airflow/jobs/local_task_job.py:25]: added.

File: airflow/jobs/local_task_job.py (hunk 4 above)
Comment (milton0825) [reply airflow/jobs/local_task_job.py]: fixed.

File: airflow/jobs/scheduler_job.py (hunk 5 above)
Comment (milton0825) [reply airflow/jobs/scheduler_job.py]: fixed.

File: airflow/jobs/scheduler_job.py (hunk 6 above)
Comment (milton0825) [reply airflow/jobs/scheduler_job.py]: added

File: tests/jobs/test_local_task_job.py (hunk 7 above)
Comment (milton0825) [reply tests/jobs/test_local_task_job.py]: Looks like it is not used. I removed it.

File: tests/jobs/test_scheduler_job.py (hunk 9 above)
Comment (milton0825) [reply tests/jobs/test_scheduler_job.py]: Yeah that is a good idea. Fixed

File: tests/jobs/test_base_job.py (hunk 8 above)
Comment (milton0825) [reply tests/jobs/test_base_job.py]: removed.

File: tests/www/api/experimental/test_dag_runs_endpoint.py (hunk 10 above)
Comment (milton0825) [reply tests/www/api/experimental/test_dag_runs_endpoint.py:38]: Actually there is a need for this. In `test_jobs.py`, we clear the `dag` table before the tests, which includes the example dags used by other tests such as `test_dag_runs_endpoint.py`. In the past `test_jobs.py` runs after all other tests so everything looks fine.

However, after this change, the test execution order also changes and it expose the race condition here. So I added this change to reload the `dag` table in tests that requires example dags in the `dag` table.

Comment (BasPH) [reply tests/www/api/experimental/test_dag_runs_endpoint.py:38]: Clear, thanks!

**Pull request merged**